### PR TITLE
Collector main components general refactor

### PR DESCRIPTION
**Pull request by itamararjuan**
…igh memory during initial ramp up of logs

### CODE REVIEW

File: collector/app.js [hunk 1]
```diff
@@ -1,39 +1,54 @@
...
-    process.exit(1);
 }
 
-// const somePod = new Pod({
-//     targetUrl: 'http://docker.for.mac.host.internal:3030/logs/nn',
-//     serviceName: 'nn',
-// });
-// somePod.start();
+continuousSetup();
+setInterval(() => { continuousSetup(); }, reconfigIntervalInMilli);
```
Comment (ronnno) [inline collector/app.js]: here in would also be nice to use setTimeout to avoid overlapping executions

File: collector/app.js [hunk 2]
```diff
@@ -1,39 +1,54 @@
...
+const reconfigIntervalInMilli = 86400 * 1000;
+const pods = [];
+let config;
 
-if (fs.existsSync(configPath)) {
-    try {
-        const configAsString = fs.readFileSync(configPath, 'utf-8');
-        const config = JSON.parse(configAsString);
+async function continuousSetup() {
+    console.log('Configuring pods...');
```
Comment (ronnno) [inline collector/app.js]: `pods` is obscure. you can name it by thinking about what data, scope, or action it represents. it could represent a log stream collector, or service logs collector - since the interface "tells" us that each service has one stream of logs.

File: collector/app.js [hunk 3]
```diff
@@ -1,39 +1,54 @@
...
+
+const reconfigIntervalInMilli = 86400 * 1000;
+const pods = [];
+let config;
 
-if (fs.existsSync(configPath)) {
-    try {
-        const configAsString = fs.readFileSync(configPath, 'utf-8');
-        const config = JSON.parse(configAsString);
+async function continuousSetup() {
```
Comment (ronnno) [inline collector/app.js]: name is unclear. "continuous" tells us how it's being used not what it does, and "setup" can be made more explicit. 
 this method  tracks the network configuration and responds to changes in the network - by launching and registering collector processes and stopping them, right?

so it can be something like "[monitor/inspect/poll/apply]NetworkConfig" or "[supervise/launchOrDestroy]CollectorProcesses" (pods?) etc.

File: collector/app.js [hunk 4]
```diff
@@ -1,39 +1,54 @@
...
+    console.log('Configuring pods...');
+    if (!config) {
+        console.log('initial run');
+        // Initial run
+        config = await getConfiguration();
```
Comment (ronnno) [inline collector/app.js:13]: getConfiguration hides the fact that we are interrogating the network at this stage to find produce a map of available log sources. it sounds like a simple getter

File: collector/app.js [hunk 5]
```diff
@@ -1,39 +1,54 @@
...
-    } catch (e) {
-        console.log('Failed to startup collector', e);
-        process.exit(2);
+
+        stats.setup(pods);
+    } else {
+        console.log('subsequent run');
+        // Subsequent runs
+        let newConfig = await getConfiguration();
+        config = mergeConfigs({
```
Comment (ronnno) [inline collector/app.js]: - the name `mergeConfigs` would work if this was a class with a state variable config. as it is it's comparing or detects changes. it does not merge.
- it should not return the merged config because it is identical to newConfig

File: collector/app.js [hunk 6]
```diff
@@ -1,39 +1,54 @@
...
+const pods = [];
+let config;
 
-if (fs.existsSync(configPath)) {
-    try {
-        const configAsString = fs.readFileSync(configPath, 'utf-8');
-        const config = JSON.parse(configAsString);
+async function continuousSetup() {
+    console.log('Configuring pods...');
+    if (!config) {
```
Comment (ronnno) [inline collector/app.js:11]: what needs to change in your implementation so that this if can be removed and only the `else` section stays?

File: collector/app.js [hunk 7]
```diff
@@ -1,39 +1,54 @@
...
+setInterval(() => { continuousSetup(); }, reconfigIntervalInMilli);
 
-// pods.push(somePod);
+// process.on('unhandledRejection', () => {
+//     // For now we just want unhandled rejections to not output to console.
+// });
 
-stats.setup(pods);
\ No newline at end of file
+process.on('uncaughtException', () => {
```
Comment (ronnno) [inline collector/app.js:51]: why not log these exceptions?

File: collector/app.js [hunk 8]
```diff
@@ -1,39 +1,54 @@
...
-// const somePod = new Pod({
-//     targetUrl: 'http://docker.for.mac.host.internal:3030/logs/nn',
-//     serviceName: 'nn',
-// });
-// somePod.start();
+continuousSetup();
+setInterval(() => { continuousSetup(); }, reconfigIntervalInMilli);
 
-// pods.push(somePod);
+// process.on('unhandledRejection', () => {
```
Comment (ronnno) [inline collector/app.js:47]: delete or uncomment this

File: collector/library/config.js [hunk 9]
```diff
@@ -3,16 +3,10 @@
...
-const { promisify } = require('util');
-const writeFile = promisify(fs.writeFile);
-
 const fetch = require('node-fetch');
 
-console.log('Querying the current state of the nodes on the Orbs Network V2...');
-
-(async function () {
+async function getConfiguration() {
+    console.log('Querying the current state of the nodes on the Orbs Network V2...');
```
Comment (ronnno) [inline collector/library/config.js:9]: some of these words (query, network for example) should be in the function name...

File: collector/library/config.js [hunk 10]
```diff
@@ -36,12 +31,42 @@ console.log('Querying the current state of the nodes on the Orbs Network V2...')
...
+
+    for (let n in oldConfig) {
+        const podConfig = oldConfig[n];
+        if (newConfig.findIndex(c => c.targetUrl === podConfig.targetUrl) === -1) {
+            // This is a pod that we do not need to monitor anymore.
+            removeCallback(podConfig);
+        }
+    }
+
+    return mergedConfig;
```
Comment (ronnno) [inline collector/library/config.js]: this always holds: `mergedConfig == newConfig` so no need to return and possibly no need to declare it

File: collector/library/pod.js [hunk 11]
```diff
@@ -16,19 +16,22 @@ const readFile = util.promisify(fs.readFile);
 /**
```
Comment (ronnno) [inline collector/library/pod.js:15]: lets go over all todos in the project and remove anything you will not do now

File: collector/library/pod.js [hunk 12]
```diff
@@ -16,19 +16,22 @@ const readFile = util.promisify(fs.readFile);
...
- *    without creating timers for each and every packet of data. 
+ *    without creating timers for each and every packet of data.
  * 
  *    As in, if the first packet sending to Logstash fails, abort the entire process and wait for the next
  *    file checker iteration in order to send the data out.
  * 
  *    Leave the mechanism for re-trying packets aside for now (let's see if this improves memory performance somehow)
  * 
+ * 5. Persistent mechanism & Startup review together with Ron
+ * 6. Reconfigure live while the process is running
```
Comment (ronnno) [inline collector/library/pod.js]: isnt this done?

File: collector/library/pod.js [hunk 13]
```diff
@@ -46,7 +49,7 @@ class Pod {
 
         this.serviceName = serviceName;
         this.targetUrl = targetUrl;
-        this.checkInterval = checkInterval;
+        this.checkIntervalInMilli = checkInterval;
```
Comment (ronnno) [inline collector/library/pod.js]: name should say what is being checked / done.

Comment (ronnno) [inline collector/library/pod.js]: lets discuss why there is an interval here in the first place - unless there is a network error or a cut connection we should never sleep - once the batches are downloaded - there should normally always be a new batch to start tailing. this interval is really only relevant for slowing down in case there is a persistent network problem. so this really is a retry interval. and we should fix it so there is no unneeded waiting... am i correct?

File: collector/library/pod.js [hunk 14]
```diff
@@ -66,72 +69,72 @@ class Pod {
     }
 
     async checkForNewFiles() {
-        console.log(`[START][${this.state}] periodical check for new files`);
-        if (this.state === 'disabled') {
-            return;
-        }
+        console.log(`[START] periodical check for new files`, this.targetUrl);
```
Comment (ronnno) [inline collector/library/pod.js]: - files means batches? lets keep the language uniform
- this method does not just check if there are new batches to download them. it is actually downloading them them eagerly until a network error, exception or something happens in the collector logic to stop it. (like a SIGTERM for instance)

File: collector/library/pod.js [hunk 15]
```diff
@@ -66,72 +69,72 @@ class Pod {
...
+                    console.log('could not download the entire batch', batch);
+                    return;
+                } // Avoid moving to next batch in case we could not download the entire batch
             }
         }
     }
 
+    stop() {
+        console.log(`Shutting down pod for endpoint ${this.targetUrl}`);
+        clearInterval(this.pid);
```
Comment (ronnno) [inline collector/library/pod.js:114]: this can lead to race conditions - its possible that right after this clearing we arrive at the finally below and get a new pid

**Review by ronnno — COMMENTED** [review]
still did not review `handleBatch`

File: collector/app.js (hunk 1 above)
Comment (itamararjuan) [reply collector/app.js]: The interval is once a day, do you think there is a chance they might overlap over a single complete day?

File: collector/app.js (hunk 2 above)
Comment (itamararjuan) [reply collector/app.js]: The name Pod is just a figment of my imagination. It's basically a monitoring unit per URL. I just couldn't find a good enough name for it. I agree Pod is obscure. What ideas you had in mind to replace this terminology with?

File: collector/app.js (hunk 3 above)
Comment (itamararjuan) [reply collector/app.js]: hmm maybe update & apply ? I agree continuous is not the right wording. I suggest calling it `syncNetworkConfig` maybe?

File: collector/app.js (hunk 4 above)
Comment (itamararjuan) [reply collector/app.js:13]: `getConfiguration` gets the actual network sources as they are now without any interrogations taking place. The interrogation happens in this code inline. Why do you think it hides it then?

File: collector/app.js (hunk 5 above)
Comment (itamararjuan) [reply collector/app.js]: * I propose we rename it to: `detectConfigChanges` ?
* Agreed - to be changed.

File: collector/library/pod.js (hunk 13 above)
Comment (itamararjuan) [reply collector/library/pod.js]: This is the time in ms which we will sleep between attempts to handle data from our designated source.
It can be used however in multiple failing scenarios (either logstash not ready, we failed downloading the batch, the endpoint hasn't updated to a boyar with the logs-service)

I want to call it just generally `timeoutInMs` but not sure it's the best  name. I would love your help to try and nail the correct name for this specific number.

Comment (itamararjuan) [reply collector/library/pod.js]: We basically don't want to hammer logstash, not drain our network connections and basically provide some kind of spacing between the other pods competing for resources. It can maybe be lowered but I do think it makes sense in having it. Since we "retry" for various and diverse reasons - it makes a lot of sense that it's here.
The only case, in my opinion where it's not really necessary is when we successfully finished inserting a batch and we can easily move to the next batch. In all other cases of errors, you'd want to allow spacing between attempts and let the other systems (logstash and the logs server) some breathing room. That is my rubber ducking to why I think it's needed here.

File: collector/library/pod.js (hunk 14 above)
Comment (itamararjuan) [reply collector/library/pod.js]: * yes - files means batches (I'll change the wording)
* The name of the method neesd to change. It historically used to emit events about new batches to the events bus. We should probably rename it to something along the lines of: `collect` ?

File: collector/library/pod.js (hunk 15 above)
Comment (itamararjuan) [reply collector/library/pod.js:114]: Good catch!
I'll make sure to set a property on the instance of the Pod. Checking it in the finally before I set the new pid.
That should do the trick I think.

File: collector/library/config.js [hunk 16]
```diff
@@ -39,7 +39,7 @@ async function getConfiguration() {
         }
     }
 
-    return pods;
+    return pods.slice(0, 25);
```
Comment (ronnno) [inline collector/library/config.js]: ?

File: collector/library/pod.js [hunk 17]
```diff
@@ -125,10 +112,12 @@ class Pod {
         await new Promise((r) => setTimeout(r, (Math.floor(Math.random() * 30 * 1000))));
         await this.createWorkDir();
 
-        let counter = 1;
-
         const setupTimer = () => {
             this.checkForNewFiles()
+                .catch((_) => {
+                    // Do nothing
+                    console.log(_);
```
Comment (ronnno) [inline collector/library/pod.js]: ?

File: collector/library/pod.js [hunk 18]
```diff
@@ -224,15 +210,57 @@ class Pod {
      * 
      */
     async submitDataToLogstash({ data, batch, }) {
-        const body = JSON.stringify({ message: data.toString() });
+        const dataAsString = data.toString();
+        let lines;
+
+        if (this.remainder.length > 0) {
```
Comment (ronnno) [inline collector/library/pod.js]: this if is redundant. you can always add remainder wether it's empty or not. just watch out for undefined...

File: collector/library/pod.js [hunk 19]
```diff
@@ -224,15 +210,57 @@ class Pod {
...
+                dataObject._collector_batchId = batch.id;
+                dataObject._collector_source_url = this.targetUrl;
+                body = JSON.stringify(dataObject);
+            } else {
+                // From here we can have one of 2 options:
+                // 1. a partial JSON log line
+                // 2. a plain text log
+
+                if (i === lines.length - 1) { // This is the last line from []lines
+                    // This is a partial json, carry this remainder over to the next iteration of submitDataToLogstash()
```
Comment (ronnno) [inline collector/library/pod.js]: not accurate comment - even if it's partial it could still be text. the code would work just the same

File: collector/library/pod.js [hunk 20]
```diff
@@ -224,15 +210,57 @@ class Pod {
...
+        } else {
+            lines = dataAsString.split('\n');
+        }
 
-        await fetch('http://logstash:5000/', {
-            method: 'post',
-            body,
-            headers: { 'Content-Type': 'application/json' },
-        });
+        for (let i = 0; i < lines.length; i++) {
```
Comment (ronnno) [inline collector/library/pod.js]: simplify the flow 

```
for (let i=0; i < lines.length - 1; i++) { // we always skip the last line - it is for reminder
...
}
this.remainder = lines[lines.length - 1];

File: collector/library/pod.js [hunk 21]
```diff
@@ -224,15 +210,57 @@ class Pod {
...
+            const line = lines[i];
+            let body;
+
+            if (this.isJSON(trim(line))) {
```
Comment (ronnno) [inline collector/library/pod.js]: instead of asking if it's json or not - insert the wrapping into `this.isJSON()` so that you don't need a condition:
```
const dataObject = this.jsonify(line);
dataObject._collector_batchId = batch.id;
dataObject._collector_source_url = this.targetUrl;
body = JSON.stringify(dataObject);
```

File: collector/library/pod.js [hunk 22]
```diff
@@ -224,15 +210,57 @@ class Pod {
...
+                const dataObject = JSON.parse(trim(line));
+                dataObject._collector_batchId = batch.id;
+                dataObject._collector_source_url = this.targetUrl;
+                body = JSON.stringify(dataObject);
+            } else {
+                // From here we can have one of 2 options:
+                // 1. a partial JSON log line
+                // 2. a plain text log
+
+                if (i === lines.length - 1) { // This is the last line from []lines
```
Comment (ronnno) [inline collector/library/pod.js]: for loop should never enter for the last element - see comment above

File: collector/library/pod.js [hunk 23]
```diff
@@ -224,15 +210,57 @@ class Pod {
...
+                        textMessage: line,
+                    };
+
+                    dataObject._collector_batchId = batch.id;
+                    dataObject._collector_source_url = this.targetUrl;
+                    body = JSON.stringify(dataObject);
+                }
+            }
+
+            if (body !== undefined && this.isJSON(body) > 0) {
```
Comment (ronnno) [inline collector/library/pod.js]: body is always JSON - you made sure of that - why parse it again?

File: collector/library/pod.js [hunk 24]
```diff
@@ -224,15 +210,57 @@ class Pod {
...
+                await fetch('http://logstash:5000/', {
+                    method: 'post',
+                    body,
+                    headers: { 'Content-Type': 'application/json' },
+                });
 
-        await this.documentSentData({ data, batch });
+                await this.documentSentData({ data: Buffer.from(line, 'utf-8'), batch });
```
Comment (ronnno) [inline collector/library/pod.js]: i think you are forgetting to count your line feeds here...

File: collector/library/stats.js [hunk 25]
```diff
@@ -17,24 +19,25 @@ const stats = {
...
 
-            if (!logstashConnected) {
-                logstashConnected = !pod.logstashConnectivityReported;
-            }
-
             totalSentBytes += pod.stats.totalSentBytes;
             totalUnackedBytes += pod.stats.totalUnackedBytes;
         }
 
+        if (totalSentBytes > lastSentBytes){
```
Comment (ronnno) [inline collector/library/stats.js:24]: what are we doing with `totalUnackedBytes`? this number cannot be interesting because we never send more than one line - if it fails, we start the entire flow anew. if it succeeds this number comes back to zero. so it will always be some small value or zero - consider deleting this variable from memory and avoid printing it

File: collector/library/pod.js (hunk 14 above)
Comment (itamararjuan) [reply collector/library/pod.js]: Decided to call it for now `checkAndProcessNewBatches` - we can discuss this further for v1.1 :-)

File: collector/library/pod.js (hunk 20 above)
Comment (itamararjuan) [reply collector/library/pod.js]: Any concrete suggestions here? I cannot think of anything in particular

File: collector/library/pod.js (hunk 21 above)
Comment (itamararjuan) [reply collector/library/pod.js]: Couldn't follow what you mean

**Comment by itamararjuan** [comment]
Code is ready for final review after the modifications discussed in recent days :-)

File: collector/library/pod.js [hunk 26]
```diff
@@ -162,6 +165,11 @@ class Pod {
 
         await new Promise((res, rej) => {
             request.body.on('data', async (data) => {
+                if (this.dead === true) { // Our pod has stopped, kill this request
```
Comment (ronnno) [inline collector/library/pod.js]: pods that are currently not receiving any data will never check for `dead`, this should be an event or ride on something else like close the incoming connection etc.

File: collector/library/pod.js [hunk 27]
```diff
@@ -156,46 +154,72 @@ class Pod {
...
+                request.body.destroy();
+            });
 
-            } catch (err) {
-                // The request got an error for some reason,
-                // The next round for new files scouting will pick up where we left off ;-)
-            }
-        });
+            request.body.on('data', async (data) => {
+                if (this.dead === true) { // Our pod has stopped, kill this request
```
Comment (ronnno) [inline collector/library/pod.js]: this is no longer needed. 

Notice, that if you want to guarantee that we never send any new bytes after a "pod" is killed, then you have a bug here because you just close the promise, destroy the request body, and continue to handle the incoming bytes ignoring the dead flag.

File: collector/library/pod.js [hunk 28]
```diff
@@ -156,46 +154,72 @@ class Pod {
...
+                    });
+                }
+
+                bytesSentAccumlator += data.length;
+            });
 
-        this.em.on(UNACKED_DATA, async ({ data, batch, packetId }) => {
-            await this.submitDataToLogstash({ data, batch, packetId });
+            request.body.on('error', (err) => {
+                console.log(err);
```
Comment (ronnno) [inline collector/library/pod.js:203]: I know I said before that we might want to ignore errors but now I don't see the problem with killing the request object and throwing an exception

File: collector/library/pod.js [hunk 29]
```diff
@@ -208,52 +232,30 @@ class Pod {
...
-        const body = JSON.stringify({ message: data.toString() });
+    async submitDataToLogstash({ data, batch, }) {
+        const dataAsString = data.toString();
+        let lines;
+
+        lines = (this.remainder + dataAsString).split('\n');
+        this.remainder = '';
+
+        // Run on all lines but the last one!
+        for (let i = 0; i < lines.length - 1; i++) {
```
Comment (ronnno) [inline collector/library/pod.js:246]: since you have no good way of distinguishing between a connection that was closed because of a server error, gateway / proxy issue or otherwise, i now think it's a real bug to move on to the next file after the connection is closed.
the correct behavior after a connection is closed is to start again from the top of the method - we need to get the manifest, compare the number of bytes handled for any batch, find the current batch to download, and download only it. then next iteration we must download the manifest again,
in short, this loop here is a bug. you should only look for the first file that was never entirely retrieved and download only it.

**Review by ronnno — APPROVED** [review]
Several of these comments are not addressed. I don't think there is anything that should prevent deployment - so I'm approving, but lets go over all the comments here together and open issues for anything that was not addressed.
also, any todo still remaining should have an issue..
